include/rootston/output.h

```diff
@@ -0,0 +1,49 @@
+#ifndef _ROOTSTON_OUTPUT_H
+#define _ROOTSTON_OUTPUT_H
+
+#include <time.h>
+#include <pixman.h>
+#include <wayland-server.h>
+
+/**
+ * Damage tracking requires to keep track of previous frames' damage. To allow
+ * damage tracking to work with triple buffering, a history of two frames is
+ * required.
+ */
+#define ROOTS_OUTPUT_PREVIOUS_DAMAGE_LEN 2
+
+struct roots_desktop;
+
+struct roots_output {
+	struct roots_desktop *desktop;
+	struct wlr_output *wlr_output;
+	struct wl_list link; // roots_desktop:outputs
+
+	struct roots_view *fullscreen_view;
+
+	struct timespec last_frame;
+	pixman_region32_t damage; // in ouput-local coordinates
+
+	// circular queue for previous damage
+	pixman_region32_t previous_damage[ROOTS_OUTPUT_PREVIOUS_DAMAGE_LEN];
+	size_t previous_damage_idx;
```

Comment:
I don't think so. The compositor should be able to choose `ROOTS_OUTPUT_PREVIOUS_DAMAGE_LEN`, or to do not use damage tracking. Also, the previous damage needs to be applied to the current damage, that can't happen in `wlr_output`.

See my `wlr_output_damage` proposal.

Reply:
> The compositor should be able to choose ROOTS_OUTPUT_PREVIOUS_DAMAGE_LEN

Why?

Reply:
It might not want to do damage tracking, or choose an history with a greater depth.

Reply:
If it was an optional component, I don't think it would be an issue since compositors could always roll their own. But I don't think baking this directly into `wlr_output` is a good idea.

Reply:
Hmm, you're right.

Reply:
What's the resolution here?

Reply:
Please give feedback about my `wlr_output_damage` proposal

Reply:
It looks good but I do have some comments, can you spin off a ticket for it?

Reply:
#614